### PR TITLE
Replace `findNodeHandle` with `this.nativeRef` in `ReactNativeComponent`

### PR DESCRIPTION
**Pull request by yungsters**
`ReactNativeComponent` currently uses `findNodeHandle` and `mountSafeCallback`. The former is not desirable. The latter is bad because now that `View` and `Text` have been upgraded to extend `ReactNativeComponent`, `mountSafeCallback` uses `isMounted` and triggers the React warning.

This resolves the issue by requiring that subclasses assign `this.nativeRef` to the root element's `ref` prop.

### CODE REVIEW

File: packages/react-native-renderer/src/ReactNativeComponent.js [hunk 1]
```diff
@@ -22,7 +22,6 @@ import TextInputState from 'TextInputState';
 import UIManager from 'UIManager';
 
 import * as ReactNativeAttributePayload from './ReactNativeAttributePayload';
-import {mountSafeCallback} from './NativeMethodsMixinUtils';
 
 export default function(
   findNodeHandle: any => ?number,
```
Comment (sophiebits) [inline packages/react-native-renderer/src/ReactNativeComponent.js]: rm?

**Comment by sebmarkbage** [comment]
The proper solution is supposed to be to use forwardRefs and not ReactNativeComponent.

File: packages/react-native-renderer/src/ReactNativeComponent.js [hunk 2]
```diff
@@ -37,36 +36,48 @@ export default function(
...
      */
     props: Props;
     state: State;
 
+    /**
+     * Subclasses should set the `ref` prop on the root element to this.
+     */
+    nativeRef = (nativeElementRef: ?any): void => {
+      this._nativeTag =
+        nativeElementRef != null ? nativeElementRef._nativeTag : null;
```
Comment (sebmarkbage) [inline packages/react-native-renderer/src/ReactNativeComponent.js]: `_nativeTag` is a private implementation detail of the stateNode in this particular renderer. I want to change it in Fabric. I also wanted it to be lazy and this forces it to be eager. You're better off storing the actual ref and then call `findNodeHandle` on it.

**Comment by sebmarkbage** [comment]
This will still be an issue. https://github.com/facebook/react/pull/12747/files#diff-3899566ffae25dd959157d1e7a9fc8c1R162

**Comment by yungsters** [comment]
You’re right. I think I can just use `forwardRef` instead of this now. Will try it out tomorrow. Thanks!

**Comment by sebmarkbage** [comment]
I think the issue could be if your native components are adding additional public APIs on top. And ofc forwardRefs is a bit of a pain for classes since you have to spread the ref onto the props.

**Comment by yungsters** [comment]
I'm going to instead deprecate `ReactNative.NativeComponent` in favor of `React.forwardRef` as suggested by @sebmarkbage.

**Pull request closed without merging**